text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,660 @@
+  +------------+          .     +---------+------------+
+  | Publish m2 |. . . . . .               | Process m1 |
+  +------------+                          +------------+------------+
+                                                       | Process m2 |
+                                                       +------------+
+```
+
+#### Single message producers, batch pull-based consumer with manual settlement
+
+```
+  PRODUCER                    CONSUMER
+
+  +------------+
+  | Publish m1 |. . . .       +----------------------------------------+
+  +------------+      .       | Ambient                                |
+                      .       +-+---------+----------------------------+
+                      . . . . . | Receive |
+  +------------+          .     +---------+    +-----------+
+  | Publish m2 |. . . . . .                    | Settle m1 |
+  +------------+                               +-----------+-----------+
+                                                           | Settle m2 |
+                                                           +-----------+
+```
+
+## Future possibilities
+
+### Transport context propagation
+
+One possibility to seamlessly integrate producer/consumer and intermediary
+instrumentation in a flexible and extensible way would be the introduction of a
+second transport context layer in addition to the creation context layer.
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. An additional _transport context layer_ allows correlating the producer and
+   the consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries that
+   are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+This would keep the existing correlation between producers and consumer intact,
+while allowing intermediaries to use the transport context to correlate
+intermediary instrumentation with existing producer and consumer
+instrumentations.
+
+### Standards for context propagation
+
+Currently, instrumentors have to decide how to attach and extract context from
+messages in order to fulfill the [requirements for context propagation](#context-propagation).
+While preserving the freedom for instrumentor to choose how to propagate
+context, in the future these conventions should list recommended ways of how to
+propagate context using popular messaging protocols.
```

Comment:
Hi, just leaving a comment as I came across [an issue](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1202) that can be resolved if some sort of standards are in place.

If you refer to the issue above, it is apparent that there is already a divergence of approaches to propagate context with different instrumentors in different languages. I wonder if more attention should be given to establishing a convention before everyone starts doing things differently and becoming out of control.

Reply:
Hi @melvinkcx 
A while ago this OTEP https://github.com/open-telemetry/oteps/blob/main/text/trace/0205-messaging-semantic-conventions-context-propagation.md was merged to make the context propagation a bit more explicit. 

It doesn't go in exact details on how this should be done (where in the message to put the context, under which "name" etc) because there's no finalized standards for it. The OTEP only makes it clear that the context should be transported together with the message and should be immutable. Since each protocol/messaging system does something different, it's hard to come up with a specification that works for everyone. Our intentions for now is to define the minimum on how propagation will work, and leave to instrumentation to implement it using their appropriate idioms. For ex, this draft [document for AMQP](https://w3c.github.io/trace-context-amqp/#trace-context-fields-placement-in-a-message) tells to add the context in `application-properties`. 

Of course, once stable specifications are available, we will update the recommendations in the OTel conventions, like the AMQP one.

Reply:
Thanks for the insights!

Reply:
>  Since each protocol/messaging system does something different, it's hard to come up with a specification that works for everyone

How about specifying just a guideline / best-practice like: "Call the propagator on the message headers or equivalent"

Reply:
I opened a PR to bring the contents of the OTP into the spec. Maybe that already gives the direction we need, without having to be precise on where to add. https://github.com/open-telemetry/opentelemetry-specification/pull/2750

Reply:
One interesting line on the W3C document for AMQP linked above:
"...AMQP message section "application-properties" is immutable collection of properties"
They recommend to use application-properties because is immutable data that brokers cannot change. This aligns with the immutable Creation Context from above. 
The MQTT doc doesn't mention immutability but defines user-properties and the method to use

Reply:
@brunobat and that's for MQTT v5 you are referring to I guess. Thinking at MQTT 3.1.1 there is no notion of application/user properties, everything is in your payload.